Fixed persistent volume claim controllers processing an old volume

jsafrane opened the pull request:
Very similar to https://github.com/kubernetes/kubernetes/pull/19868, now about volumes instead of claims. I've seen a GCE volume provisioned several times, i.e. one Kubernetes PersistentVolume created several GCE PDs.

We should always load the newest version of the volume from APIserver before
processing it.

When PersistentVolumeProvisionerController.reconcileVolume() is called with the
same volume in short succession (e.g. the volume is created by a provisioner
and at the same time periodic check of all volumes is scheduled), the second
reconcileVolume() call gets an old copy of the volume as its parameter and
it does not see annotations updated by the previous call.

This may result in one volume being provisioned several times, creating orphan
volumes in the cloud.

The same error is in PersistentVolumeClaimBinder.syncVolume().

jsafrane commented:
@kubernetes/rh-storage, @saad-ali

markturansky commented:
@jsafrane is there a way to query GCE by volume name (not ID) before we provision a new one?

I commented in your other PR that we can use the PV directly on the GCEPD.Name field itself (GCEPD.ID is GCE's unique identifier).  We already have a unique name assigned by the API server.

jsafrane commented:
Yes, there is a way to query GCE volumes by Name. It would need to be implemented in all dynamic provisioning plugins + I am not sure it would prevent all the race conditions here.

markturansky commented on pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go:
```diff
@@ -193,6 +193,14 @@ func (binder *PersistentVolumeClaimBinder) deleteClaim(obj interface{}) {
+	if err != nil {
+		return fmt.Errorf("Cannot reload volume %s: %v", volume.Name, err)
+	}
+	volume = newPv
```
per our IRC discussion, please update the local cache with the latest and greatest version.

jsafrane replied:
Added (few lines below).

markturansky replied:
We use the volumeIndex custom Store impl because we needed fancy indexing and sorting.  The watch still has its own store.

We would need to add the PV to both.  

Ideally, we allow the Controller framework to take a store argument it uses for watch data.  It does not currently, so I couldn't pass our volumeIndex to it.  This controller doesn't need a 2nd store. We need to be able to specify the one we want to use in the watch.

markturansky replied:
Nevermind, I am confusing my controllers. There is no local store use in this one. We use the volumeIndex custom store only.

jsafrane committed:
Fixed persistent volume claim controllers processing an old volume

We should always load the newest version of the volume from APIserver before
processing it.

When PersistentVolumeProvisionerController.reconcileVolume() is called with the
same volume in short succession (e.g. the volume is created by a provisioner
and at the same time periodic check of all volumes is scheduled), the second
reconcileVolume() call gets an old copy of the volume as its parameter and
it does not see annotations updated by the previous call.

This may result in one volume being provisioned several times, creating orphan
volumes in the cloud.

The same error is in PersistentVolumeClaimBinder.syncVolume().

swagiaal commented:
Same reservations as #19868...  LGTM though
@markturansky

jsafrane commented:
@k8s-bot test this issue: #21029
